File: apps/mobile/src/modules/photos/profileStats.ts
export interface ProfileStatsPhoto {
  dateTaken: string | null
  camera: string | null
  lens: string | null
}

export interface ProfileStats {
  photoCount: number
  cameraCount: number
  lensCount: number
  yearSpan: string | null
}

export function collectProfileStats(photos: ProfileStatsPhoto[]): ProfileStats {
  const cameras = new Set<string>()
  const lenses = new Set<string>()
  let minYear = Number.POSITIVE_INFINITY
  let maxYear = Number.NEGATIVE_INFINITY

  for (const photo of photos) {
    if (photo.camera) {
      cameras.add(photo.camera)
    }
    if (photo.lens) {
      lenses.add(photo.lens)
    }
    if (photo.dateTaken) {
      const year = new Date(photo.dateTaken).getFullYear()
      if (!Number.isNaN(year)) {
        minYear = Math.min(minYear, year)
        maxYear = Math.max(maxYear, year)
      }
    }
  }

  const yearSpan = Number.isFinite(minYear) ? (minYear === maxYear ? String(minYear) : `${minYear}–${maxYear}`) : null

  return {
    photoCount: photos.length,
    cameraCount: cameras.size,
    lensCount: lenses.size,
    yearSpan,
  }
}

File: apps/mobile/Afilmory/Core/Models/ProfileStats.swift
import Foundation

struct ProfileStats: Codable, Equatable, Sendable {
  let photoCount: Int
  let cameraCount: Int
  let lensCount: Int
  let yearSpan: String?

  static func collect(_ photos: [GalleryPhoto], timeZone: TimeZone = .current) -> ProfileStats {
    var cameras = Set<String>()
    var lenses = Set<String>()
    var years: [Int] = []
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    for photo in photos {
      if let camera = photo.camera { cameras.insert(camera) }
      if let lens = photo.lens { lenses.insert(lens) }
      if let value = photo.dateTaken, let date = PhotoDateParser.date(value, timeZone: timeZone) {
        years.append(calendar.component(.year, from: date))
      }
    }
    let minimum = years.min()
    let maximum = years.max()
    let yearSpan: String?
    if let minimum, let maximum {
      yearSpan = minimum == maximum ? String(minimum) : "\(minimum)–\(maximum)"
    } else {
      yearSpan = nil
    }
    return ProfileStats(
      photoCount: photos.count,
      cameraCount: cameras.count,
      lensCount: lenses.count,
      yearSpan: yearSpan
    )
  }
}
